pkg/backup/backup_compaction.go

```diff
@@ -0,0 +1,613 @@
+	tasks = append(tasks, func(ctx context.Context) error {
+		return genSpan(ctx, spanCh)
+	})
+	tasks = append(tasks, func(ctx context.Context) error {
+		return runCompaction(ctx, execCtx, encryption, spanCh, newDetails, backupManifest, progCh, store)
+	})
+	tasks = append(tasks, func(ctx context.Context) error {
+		processProg(ctx, progCh, backupManifest)
+		return nil
+	})
+
+	if err := ctxgroup.GoAndWait(ctx, tasks...); err != nil {
+		return err
+	}
+	return concludeBackupCompaction(ctx, execCtx, store, encryption, kmsEnv, backupManifest)
+}
+
+func runCompaction(
+	ctx context.Context,
+	execCtx sql.JobExecContext,
+	encryption *jobspb.BackupEncryptionOptions,
+	entries chan execinfrapb.RestoreSpanEntry,
+	details jobspb.BackupDetails,
+	manifest *backuppb.BackupManifest,
+	progCh chan execinfrapb.RemoteProducerMetadata_BulkProcessorProgress,
+	store cloud.ExternalStorage,
+) error {
+	defer close(progCh)
+	var encryptionOptions *kvpb.FileEncryptionOptions
+	if encryption != nil {
+		encryptionOptions = &kvpb.FileEncryptionOptions{Key: encryption.Key}
+	}
+	sinkConf := backupsink.SSTSinkConf{
+		ID:        execCtx.ExecCfg().DistSQLSrv.NodeID.SQLInstanceID(),
+		Enc:       encryptionOptions,
+		ProgCh:    progCh,
+		Settings:  &execCtx.ExecCfg().Settings.SV,
+		ElideMode: manifest.ElidedPrefix,
+	}
+	sink, err := backupsink.MakeSSTSinkKeyWriter(sinkConf, store, nil)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := sink.Flush(ctx); err != nil {
+			log.Warningf(ctx, "failed to flush sink: %v", err)
+			logClose(ctx, sink, "SST sink")
+		}
+	}()
+
+	for {
+		select {
+		case entry, ok := <-entries:
+			if !ok {
+				return sink.Flush(ctx)
+			}
+
+			sstIter, err := openSSTs(ctx, execCtx, entry, encryptionOptions, details)
+			if err != nil {
+				return errors.Wrap(err, "opening SSTs")
+			}
+
+			if err := processSpanEntry(ctx, sstIter, sink); err != nil {
+				return errors.Wrap(err, "processing span entry")
+			}
+		}
+	}
+}
+
+func processSpanEntry(
+	ctx context.Context, sstIter mergedSST, sink *backupsink.SSTSinkKeyWriter,
+) error {
+	defer sstIter.cleanup()
+	entry := sstIter.entry
+	startKey, endKey := storage.MVCCKey{Key: entry.Span.Key}, storage.MVCCKey{Key: entry.Span.EndKey}
+	prefix, err := backupsink.ElidedPrefix(entry.Span.Key, entry.ElidedPrefix)
+	if err != nil {
+		return err
+	}
+	if prefix != nil {
+		startKey.Key = bytes.TrimPrefix(startKey.Key, prefix)
+		endKey.Key = bytes.TrimPrefix(endKey.Key, prefix)
+	}
+	iter := sstIter.iter
+	if err := sink.Reset(ctx, sstIter.entry.Span); err != nil {
+		return err
+	}
+	scratch := make([]byte, 0, len(prefix))
+	scratch = append(scratch, prefix...)
+	for iter.SeekGE(startKey); ; iter.NextKey() {
+		var key storage.MVCCKey
+		if ok, err := iter.Valid(); err != nil {
+			return err
+		} else if key = iter.UnsafeKey(); !ok || !key.Less(endKey) {
+			break
+		}
+		value, err := iter.UnsafeValue()
+		if err != nil {
+			return err
+		}
+		fullKey := key
```

Comment:
could we operate directly on key.key instead of creating a new local var `fullKey`?

Reply:
Oh yup, for some reason I thought this would clobber something but nope it certainly doesn't